test/Transforms/Flattening/cleanup_concreteAsRoot.llzk

```diff
@@ -0,0 +1,535 @@
+// RUN: llzk-opt -split-input-file -verify-diagnostics --pass-pipeline='builtin.module(llzk-flatten{cleanup=concrete-as-root})' %s | FileCheck --enable-var-scope %s
+
+module attributes {veridise.lang = "llzk"} {
+  global.def @g : !struct.type<@HasUseFromGlobal>
+
+  function.def @free(%a: !struct.type<@HasUseFromFreeFunc>) {
+    function.return
+  }
+
+  // always preserved
+  struct.def @Main {
+    struct.field @f : !struct.type<@HasUseFromMain>
+    function.def @compute() -> !struct.type<@Main> {
+      %self = struct.new : !struct.type<@Main>
+      function.return %self : !struct.type<@Main>
+    }
+    function.def @constrain(%arg0: !struct.type<@Main>) {
+      function.return
+    }
+  }
+
+  // preserved
+  struct.def @HasUseFromGlobal {
+    function.def @compute() -> !struct.type<@HasUseFromGlobal> {
+      %self = struct.new : !struct.type<@HasUseFromGlobal>
+      function.return %self : !struct.type<@HasUseFromGlobal>
+    }
+    function.def @constrain(%arg0: !struct.type<@HasUseFromGlobal>) {
+      function.return
+    }
+  }
+
+  // preserved
+  struct.def @HasUseFromMain {
+    function.def @compute() -> !struct.type<@HasUseFromMain> {
+      %self = struct.new : !struct.type<@HasUseFromMain>
+      function.return %self : !struct.type<@HasUseFromMain>
+    }
+    function.def @constrain(%arg0: !struct.type<@HasUseFromMain>) {
+      function.return
+    }
+  }
+
+  // preserved
+  struct.def @HasUseFromFreeFunc {
+    function.def @compute() -> !struct.type<@HasUseFromFreeFunc> {
+      %self = struct.new : !struct.type<@HasUseFromFreeFunc>
+      function.return %self : !struct.type<@HasUseFromFreeFunc>
+    }
+    function.def @constrain(%arg0: !struct.type<@HasUseFromFreeFunc>) {
+      function.return
+    }
+  }
+
+  // removed (has no concrete instantiation)
+  struct.def @NoUse<[@T]> {
+    function.def @compute() -> !struct.type<@NoUse<[@T]>> {
+      %self = struct.new : !struct.type<@NoUse<[@T]>>
+      function.return %self : !struct.type<@NoUse<[@T]>>
+    }
+    function.def @constrain(%arg0: !struct.type<@NoUse<[@T]>>) {
+      function.return
+    }
+  }
+}
+// CHECK-NOT: struct.def @NoUse<[@T]> {
```

Comment:
Do you have to repeat this check or would a DAG check work?

Reply:
My understanding is that DAG must be adjacent lines. Here I'm checking against the full LLZK output but only care to check the `struct.def` lines.